### PR TITLE
DM-30373: Add butler collection-chain command line

### PR DESCRIPTION
**Pull request by timj**
## Checklist

- [ ] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

**Review by TallJimbo — APPROVED** [review]
Looks good.  If you're up for a little more work, a `--remove X` option that just does
```
children = list(registry.getCollectionChain(parent))
children.remove(X)
registry.setCollectionChain(parent, children)
```
and a `--pop` that does
```
children = registry.getCollectionChain(parent)[1:]
registry.setCollectionChain(parent, children)
```
would be super useful, too.

**Comment by timj** [comment]
@TallJimbo I've added this but you should take a look before I merge.

Did you mean `--pop` to remove the first entry in the chain? This goes against the python `list.pop()` which removes the final entry.

Rather than add a `--remove X` option I added a `--mode` option. Doing this allowed `X` to be the `CHILDREN` that he subcommand was already having to handle and meant that I could have an `extend` mode as well as a `remove` mode.

File: python/lsst/daf/butler/cli/cmd/commands.py
Comment (timj) [inline python/lsst/daf/butler/cli/cmd/commands.py]: @TallJimbo Could also do this as `--mode=pop` and let `CHILDREN` be given as integer indexes into the chain (so as to match `list.pop`) -- defaulting to pop off the end.

File: python/lsst/daf/butler/cli/cmd/commands.py
Comment (timj) [inline python/lsst/daf/butler/cli/cmd/commands.py:520]: Would `--mode=prepend` be useful?

Comment (TallJimbo) [reply python/lsst/daf/butler/cli/cmd/commands.py:520]: Yes, and more useful than `extend`, I think.  We may represent these things as Sequences in Python, but conceptually we (`pipetask` in particular) really treat them as LIFO stacks, where all of the mutations happen at the front (yeah, maybe I should have implemented them as reversed-lists in Python, but I figured ordering the container according to the search order is more important).

So I really did mean for `--pop` to act on the front, and I really only care that we provide ways to act on the front in general, but I could easily believe that also providing ways to act on the end is the best way to make it clear what all of the actions actually do.

File: python/lsst/daf/butler/cli/cmd/commands.py
Comment (TallJimbo) [reply python/lsst/daf/butler/cli/cmd/commands.py]: (see other thread - I think the difference here is that I don't think popping off the end is a good default)

Comment (timj) [reply python/lsst/daf/butler/cli/cmd/commands.py]: In Perl you `shift` to move things off the front of the list and `pop` to remove things from the end. I do worry that calling this `pop` goes against the default python definition of pop. Maybe something like:

```
$ butler collection-chain REPO --mode=pop PARENT index index index
```
would work where we default `index` to `-1`? (and since this is using `CHILDREN` allow multiple indices).

Comment (TallJimbo) [reply python/lsst/daf/butler/cli/cmd/commands.py]: How about `--mode=pop --side=front`, or `--mode=popfront`?

Comment (timj) [reply python/lsst/daf/butler/cli/cmd/commands.py]: I've just implemented the `--mode=pop` defaulting to element 0 and allowing `--mode=pop -1` to remove from the end.

Comment (timj) [reply python/lsst/daf/butler/cli/cmd/commands.py]: The main problem actually is that `-1` is an option so you have to do `-- -1` to make it work. From what you say that's not going to be a common problem and the positive integer always works as desired.

**Pull request merged**
